### PR TITLE
Define schema for AppCatalog CRD

### PR DESCRIPTION
**Pull request by sslavic**
Towards giantswarm/giantswarm#5407

### CODE REVIEW

File: pkg/apis/application/v1alpha1/app_catalog_types.go [hunk 1]
```diff
@@ -1,10 +1,106 @@
 package v1alpha1
 
 import (
+	"encoding/json"
+
 	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+const appCatalogCRDJSON = `
```
Comment (kopiczko) [inline pkg/apis/application/v1alpha1/app_catalog_types.go]: Why not yaml?

**Review by rossf7 — CHANGES_REQUESTED** [review]
What I care about here is maintainability.  This JSON will be hard to maintain and is a step backwards from what we have now IMO.

I would like us to take a step back and just add the OpenAPI validation rules as YAML. That is what this issue is about. I agree that the Go structs for the rules are difficult to work with.

If we want to change how we manage CRDs that should be a separate issue and we should have consensus in sig-operator.

**Comment by kopiczko** [comment]
I'm personally fine with defining whole CRD in the const. I'm even biased that this the way to go ATM. But choosing JSON as a format is controversial.

Having that in a separate issue/PR makes sense though.

**Comment by rossf7** [comment]
Yes my main point here is we should make smaller iterations and have consensus before making major changes. This started out as a SWH task to add some validation rules and has escalated a lot. 

I also don't think there is consensus on using JSON for this.

File: pkg/apis/application/v1alpha1/app_catalog_types.go (hunk 1 above)
Comment (sslavic) [reply pkg/apis/application/v1alpha1/app_catalog_types.go]: There are no yaml tags in the types, only json.
More details on slack thread: https://gigantic.slack.com/archives/C2MS2VB37/p1564049998054900

Comment (kopiczko) [reply pkg/apis/application/v1alpha1/app_catalog_types.go]: Ack. Yes that is a problem to some degree. We'd probably have to unmarshal to `map[string]interface{}` then marshal to json and then unmarshal to struct.

**Comment by rossf7** [comment]
@sslavic I think this can be closed?

**Comment by sslavic** [comment]
@rossf7 we discussed yesterday on Slack, options were
- to delete this PR/branch and created new one, or
- use this PR/branch, rebase it and overwrite the commit with new one that's aligned with chart CRD 

I'm fine with either, @tomahawk28 suggested he will chose latter (see https://gigantic.slack.com/archives/C76JX6YLQ/p1567415625013000?thread_ts=1567415386.011000&cid=C76JX6YLQ )

**Comment by rossf7** [comment]
@sslavic But I think the validation was added here? 😕 https://github.com/giantswarm/apiextensions/pull/280

@tomahawk28 Do you still need this PR?

**Comment by tomahawk28** [comment]
@sslavic @rossf7 Sorry, I started with another branch. We can remove this branch.

**Pull request closed without merging**